Interesting!!! the same code, I commit it, Travis CI tested failed, I reset and recommit without modifying one letter, and it builds pass. see [previous build](https://travis-ci.org/pytorch/pytorch/builds/217478648) and  [latter build](https://travis-ci.org/pytorch/pytorch/builds/217481439)
```
FAIL: test_half_tensor (__main__.TestTorch)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_torch.py", line 2653, in test_half_tensor
    self.assertEqual(x.half().float(), x, 1e-3)
  File "/home/travis/build/pytorch/pytorch/test/common.py", line 132, in assertEqual
    assertTensorsEqual(x, y)
  File "/home/travis/build/pytorch/pytorch/test/common.py", line 126, in assertTensorsEqual
    self.assertLessEqual(max_err, prec, message)
AssertionError: 0.0018029212951660156 not less than or equal to 0.001 : 
----------------------------------------------------------------------
Ran 156 tests in 14.277s
FAILED (failures=1, skipped=23)
```

--- a/test/test_nn.py
+++ b/test/test_nn.py
@@ -467,6 +467,19 @@ class TestNN(NNTestCase):
         s = nn.Sequential(l1, l2, l1, l2, subnet)
         self.assertEqual(list(s.children()), [l1, l2, subnet])
 
+    def test_dir(self):
+        linear = nn.Linear(2, 2)
+        linear._test_submodule = nn.Linear(2, 2)
+        linear._test_parameter = Parameter(torch.Tensor(2, 2))
+        linear.register_buffer('_test_buffer', torch.Tensor(2, 2))
+        keys = linear.__dir__()
+        self.assertIn('_test_submodule', keys)
+        self.assertIn('_test_parameter', keys)
+        self.assertIn('_test_buffer', keys)
+
+        for key in keys:
+            self.assertTrue(hasattr(linear, key))
+
     def test_named_children(self):
         l1 = nn.Linear(2, 2)
         l2 = nn.Linear(2, 2)

--- a/torch/nn/modules/module.py
+++ b/torch/nn/modules/module.py
@@ -461,3 +461,12 @@ class Module(object):
             tmpstr = tmpstr + '  (' + key + '): ' + modstr + '\n'
         tmpstr = tmpstr + ')'
         return tmpstr
+
+    def __dir__(self):
+        module_attrs = dir(self.__class__)
+        attrs = list(self.__dict__.keys())
+        parameters = list(self._parameters.keys())
+        modules = list(self._modules.keys())
+        buffers = list(self._buffers.keys())
+        keys = module_attrs + attrs + parameters + modules + buffers
+        return sorted(keys)

